### PR TITLE
logging and error handling improvements

### PR DESCRIPTION
**Pull request by vroldanbet**
Closes https://github.com/authzed/spicedb/issues/980
Closes https://github.com/authzed/spicedb/issues/903
Some improvements related to https://github.com/authzed/spicedb/issues/753

This PR includes various improvements around:
- handling context cancellation and deadline errors so the API returns the correct gRPC code
- reduce log noise

The commit log contains a detailed description of each change.

### CODE REVIEW

**Comment by josephschorr** [comment]
@vroldanbet Please rebase

File: internal/datastore/postgres/common/pgx.go [hunk 1]
```diff
@@ -88,6 +89,13 @@ func ConfigurePGXLogger(connConfig *pgx.ConnConfig) {
 			if level == pgx.LogLevelInfo {
 				level = pgx.LogLevelDebug
 			}
+			// do not log cancelled queries as errors
```
Comment (josephschorr) [inline internal/datastore/postgres/common/pgx.go:92]: newline before the comment (here and below)

File: internal/datastore/postgres/common/pgx.go [hunk 2]
```diff
@@ -88,6 +89,13 @@ func ConfigurePGXLogger(connConfig *pgx.ConnConfig) {
...
+			if errArg, ok := data["err"]; ok {
+				err, ok := errArg.(error)
+				if ok && (errors.Is(err, context.Canceled)) {
```
Comment (josephschorr) [inline internal/datastore/postgres/common/pgx.go:95]: are the parens necessary arounbd the `errors.Is`?

File: internal/dispatch/caching/caching.go [hunk 3]
```diff
@@ -407,10 +407,12 @@ func (cd *Dispatcher) DispatchLookupSubjects(req *v1.DispatchLookupSubjectsReque
 		for _, slice := range cachedResultRaw.([][]byte) {
 			var response v1.DispatchLookupSubjectsResponse
 			if err := response.UnmarshalVT(slice); err != nil {
-				return fmt.Errorf("could not publish cached lookup subjects result: %w", err)
+				return err
 			}
 			if err := stream.Publish(&response); err != nil {
-				return fmt.Errorf("could not publish cached lookup subjects result: %w", err)
+				// wrapping status.Status into an error means status.FromError()
```
Comment (josephschorr) [inline internal/dispatch/caching/caching.go:413]: not sure I follow...

File: internal/services/v1/errors.go [hunk 4]
```diff
@@ -269,7 +269,10 @@ func rewriteError(ctx context.Context, err error) error {
 		return status.Errorf(codes.Internal, "internal error: %s", err)
 	case errors.As(err, &graph.ErrUnimplemented{}):
 		return status.Errorf(codes.Unimplemented, "%s", err)
-
+	case errors.Is(err, context.DeadlineExceeded):
```
Comment (josephschorr) [inline internal/services/v1/errors.go:272]: Can we add a unit test for this?

File: internal/datastore/postgres/common/pgx.go (hunk 2 above)
Comment (vroldanbet) [reply internal/datastore/postgres/common/pgx.go:95]: they are not

File: internal/dispatch/caching/caching.go (hunk 3 above)
Comment (vroldanbet) [reply internal/dispatch/caching/caching.go:413]: let me try to explain it with a test-case. The assertion in this test fails:

```go
func TestStatusWontBeUnwrapped(t *testing.T) {
	err := status.Errorf(codes.NotFound, "%s", errors.New("test"))
	wrappedErr := fmt.Errorf("wrap your status error: %w", err)
	_, ok := status.FromError(wrappedErr)
	require.True(t, ok)
}

Comment (vroldanbet) [reply internal/dispatch/caching/caching.go]: another thing I explored was trying to do `errors.As` into a `status.Error` before calling `status.FromError`, but unfortunately it's internal gRPC API. Maybe you have a different idea?

File: internal/services/v1/errors.go (hunk 4 above)
Comment (vroldanbet) [reply internal/services/v1/errors.go:272]: done in https://github.com/authzed/spicedb/pull/986/commits/0c08ff8669e98ad679e6dd07023764d0f8063b4c

**Review by josephschorr — APPROVED** [review]
LGTM

**Review by josephschorr — APPROVED** [review]

**Pull request merged**